File: tools/gzip_encoder/js/gzip_encoder.js
const ENCODING = 'utf-8';
const MODE_ENCODE = 'encode';
const MODE_DECODE = 'decode';

function main(text, mode) {
  let result = ""
  if (mode === MODE_ENCODE) {
    result = gzip_encode(text);
  } else if (mode === MODE_DECODE) {
    result = gzip_decode(text);
  }
  return result;
}

function gzip_encode(text) {
  const input_bytes = new TextEncoder().encode(text);
  const gzip_bytes = pako.gzip(input_bytes);
  const base64_bytes = btoa(String.fromCharCode(...gzip_bytes));
  const output_text = base64_bytes;

  return output_text;
}

function gzip_decode(text) {
  const input_bytes = new Uint8Array(
    atob(text).split('').map((char) => char.charCodeAt(0))
  );
  const output_bytes = pako.ungzip(input_bytes);
  const output_text = new TextDecoder().decode(output_bytes);

  return output_text;
}

File: tools/gzip_encoder/gzip_encoder.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import base64
import gzip

MODE_ENCODE = "encode"
MODE_DECODE = "decode"
ENCODING = "utf-8"


def parse():
    parser = argparse.ArgumentParser()
    parser.add_argument("text", type=str)
    parser.add_argument("--mode", choices=[MODE_ENCODE, MODE_DECODE], default="encode")
    args = parser.parse_args()

    return args


def main(text: str, mode: str):
    if mode == MODE_ENCODE:
        result = gzip_encode(text)
    elif mode == MODE_DECODE:
        result = gzip_decode(text)

    return result


def gzip_encode(text: str) -> str:
    input_bytes = text.encode(ENCODING)
    gzip_bytes = gzip.compress(input_bytes)
    base64_bytes = base64.b64encode(gzip_bytes)
    output_text = base64_bytes.decode(ENCODING)

    return output_text


def gzip_decode(text: str) -> str:
    input_bytes = text.encode(ENCODING)
    base64_bytes = base64.b64decode(input_bytes)
    output_bytes = gzip.decompress(base64_bytes)
    output_text = output_bytes.decode(ENCODING)

    return output_text


if __name__ == "__main__":
    args = parse()
    result = main(args.text, args.mode)
    print(result)
